### PR TITLE
Scroll buffer after using cider-insert-X-in-repl

### PR DESCRIPTION
**Pull request by dpsutton**
As it stood, buffer would not scroll, even if you included a
`(goto-char (point-max))` call at the end. Using solution from
https://stackoverflow.com/questions/13530025/emacs-scroll-automatically-when-inserting-text

Have to watch out if the buffer is visible or not. If it is, we can
scroll it with `with-selected-window`. Otherwise, the best we can do
is the current behavior. Note `get-buffer-window` returns nil if
something is not visible which is why we need to branch on it.

## Currently

Using the `cider-insert-[last-sexp|defun|region]-inrepl` commands (C-c M-j [e,d,r]) were very convenient but would not scroll the buffer, limiting their usability.

before

Notice that the repl buffer is not scrolled. Simply adding a `(goto-char (point-max))` in `cider-insert-in-repl` was not sufficient due. (see stack overflow post from commit message above).

## After

By using `with-selected-window` we can manipulate the view of that buffer. So now we end up with this much more sensical view when using these commands:

after

### CODE REVIEW

**Comment by cichli** [comment]
We could switch to the REPL _before_ inserting to make sure the buffer is visible, but `cider-switch-to-repl-after-insert-p` would have to be renamed.

**Comment by arichiardi** [comment]
I am fine with the rename as seeing the output is more important ;)))

Thanks for opening this btw!

**Comment by dpsutton** [comment]
I've marked obsolete and created a new defcustom

```lisp
(defcustom cider-switch-to-repl-on-insert-p t
  "Whether to switch to the repl when inserting a form into the repl."
  :type 'boolean
  :group 'cider
  :package-version '(cider . "0.21.0"))

(define-obsolete-variable-alias
  'cider-switch-to-repl-after-insert-p
  'cider-switch-to-repl-on-insert-p
  "0.21.0")
```

I figure we can just say "on-insert" rather than before or after so hopefully we don't go through this again if it makes sense to focus in the buffer before/after insertion in the future.

Let me know if I didn't obsolete it correctly.

**Comment by arichiardi** [comment]
Oh it seems the problem could just be that it is not rebased on `master`.

**Comment by dpsutton** [comment]
@arichiardi do you have the correct settings? By default, CIDER doesn't eval the inserted forms automatically. use

```clojure
(setq cider-invert-insert-eval-p t)

**Comment by arichiardi** [comment]
Ah lol maybe not :)

**Comment by dpsutton** [comment]
also added fontlocking

with-fontlocking

File: cider-mode.el [hunk 1]
```diff
@@ -223,6 +223,17 @@ With a prefix argument, prompt for function to run instead of -main."
   :group 'cider
   :package-version '(cider . "0.18.0"))
 
+(defcustom cider-switch-to-repl-on-insert-p t
```
Comment (bbatsov) [inline cider-mode.el]: Defcustoms should not have predicate names IMO. I’ve never seen such naming for them.

File: cider-mode.el [hunk 2]
```diff
@@ -223,6 +223,17 @@ With a prefix argument, prompt for function to run instead of -main."
...
+  "Whether to switch to the repl when inserting a form into the repl."
+  :type 'boolean
+  :group 'cider
+  :package-version '(cider . "0.21.0"))
+
+(define-obsolete-variable-alias
```
Comment (bbatsov) [inline cider-mode.el]: I don’t see the var that was made obsolete anywhere, so I guess that’s not needed.

Comment (dpsutton) [reply cider-mode.el]: the var is right above the new one.
```lisp
(defcustom cider-switch-to-repl-after-insert-p t
  "Whether to switch to the repl after inserting a form into the repl."
  :type 'boolean
  :group 'cider
  :package-version '(cider . "0.18.0"))

(defcustom cider-switch-to-repl-on-insert-p t
  "Whether to switch to the repl when inserting a form into the repl."
  :type 'boolean
  :group 'cider
  :package-version '(cider . "0.21.0"))

(define-obsolete-variable-alias
  'cider-switch-to-repl-after-insert-p
  'cider-switch-to-repl-on-insert-p
  "0.21.0")
```

The original had a `-p` so I mimic'ed it. I'll remove that from the new one.

**Comment by arichiardi** [comment]
This looks very good to me :smile: Eager to have it in :smile: 

Thanks a lot @dpsutton!

**Comment by cichli** [comment]
Thanks for this! Sorry it's taken a while for me to get around to reviewing it.

**Comment by arichiardi** [comment]
Awesome thank you!

**Pull request merged**